crates/typst-macros/src/elem.rs

```diff
@@ -480,88 +472,904 @@ fn create_param_info(field: &Field) -> TokenStream {
+                )*
+            }
+        }
+    }
+}
+
+/// Creates the element's `Default` implementation.
+fn create_default_impl(element: &Elem) -> TokenStream {
+    let ident = &element.ident;
+    let relevant = element
+        .fields
+        .iter()
+        .filter(|field| !field.external && !field.synthesized && field.inherent())
+        .map(|Field { ident, empty, .. }| {
+            quote! { #ident: #empty }
+        });
+    let defaults = element
+        .fields
+        .iter()
+        .filter(|field| !field.external && (field.synthesized || !field.inherent()))
+        .map(|Field { ident, .. }| {
+            quote! { #ident: None }
+        });
+
+    let label_and_location = element.without_capability("Unlabellable", || {
+        quote! {
+            location: None,
+            label: None,
+            prepared: false,
+        }
+    });
+
+    quote! {
+        impl ::std::default::Default for #ident {
+            fn default() -> Self {
+                Self {
+                    span: ::typst::syntax::Span::detached(),
+                    #label_and_location
+                    guards: ::std::vec::Vec::with_capacity(0),
+                    #(#relevant,)*
+                    #(#defaults,)*
+                }
+            }
+        }
+    }
+}
+
+/// Creates the element's enum for field identifiers.
+fn create_fields_enum(element: &Elem) -> TokenStream {
+    let Elem { enum_ident, vis, .. } = element;
+
+    let mut fields = element.all_fields().collect::<Vec<_>>();
+
+    fields.sort_by_key(|field| field.forced_variant.unwrap_or(usize::MAX));
+
+    let field_names = fields.iter().map(|Field { name, .. }| name).collect::<Vec<_>>();
+    let field_consts = fields
+        .iter()
+        .map(|Field { const_ident, .. }| const_ident)
+        .collect::<Vec<_>>();
+    let field_variants = fields
+        .iter()
+        .map(|Field { enum_ident, .. }| enum_ident)
+        .collect::<Vec<_>>();
+    let definitions =
+        fields.iter().map(|Field { forced_variant, enum_ident, .. }| {
+            if let Some(variant) = forced_variant {
+                let variant = proc_macro2::Literal::u8_unsuffixed(*variant as _);
+                quote! { #enum_ident = #variant }
+            } else {
+                quote! { #enum_ident }
+            }
+        });
+
+    quote! {
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+        #[repr(u8)]
+        #vis enum #enum_ident {
+            #(#definitions,)*
+            Label = 255,
+        }
+
+        impl #enum_ident{
+            pub fn to_str(self) -> &'static str {
+                match self {
+                    #(
+                        Self::#field_variants => #field_names,
+                    )*
+                    Self::Label => "label",
+                }
+            }
+        }
+
+        impl From<#enum_ident> for u8 {
+            fn from(value: #enum_ident) -> Self {
+                value as u8
+            }
+        }
```

Comment:
As said elsewhere, I think we should probably remove this.

Reply:
It's actually necessary for the enum -> u8 conversion